src/server/endpoints/covidcast_utils/smooth_diff.py

```diff
@@ -0,0 +1,168 @@
+from collections import ChainMap
+from datetime import date
+from itertools import groupby, chain
+from numbers import Number
+from typing import Iterable, Dict, List, Optional, Union
+
+from more_itertools import windowed, peekable
+from pandas import date_range, Timedelta
+from numpy import nan, nan_to_num, array, dot, isnan
+
+
+def _smoother(values: List[Number], kernel: Optional[List[Number]] = None) -> Number:
+    """Basic smoother.
+
+    If kernel passed, uses the kernel as summation weights. If something is wrong,
+    defaults to the mean.
+    """
+    try:
+        if kernel:
+            kernel = array(kernel, copy=False)
+        else:
+            raise ValueError
+        values = array(values, copy=False)
+        smoothed_value = dot(values, kernel)
+    except (ValueError, TypeError):
+        smoothed_value = array(values).mean()
+
+    return round(smoothed_value, 8)
+
+
+def _iso_int_to_date(di: int) -> date:
+    s = str(di)
+    year, month, day = int(s[:4]), int(s[4:6]), int(s[6:8])
+    return date(year, month, day)
+
+
+def _date_to_iso_int(dt: date) -> int:
+    return int(str(dt.year) + str(dt.month).zfill(2) + str(dt.day).zfill(2))
+
+
+def _pad_group(group: Iterable[Dict], pad_length: int, pad_fill_value: Union[Number, str] = nan) -> Iterable[Dict]:
```

Comment:
have you considered using an enum instead of a string which could be anything?

Reply:
another option would be to enforce the structure of the dict (time_value, value) using TypedDicts

Reply:
made some updates to use Enums there